machine/machine_core/ssh_connection.py

```diff
@@ -30,6 +30,16 @@
 from . import timeout as timeoutlib
 
 
+def retry_write(fd, data):
```

Comment:
I would call this "write_all" and also make it handle partial writes.

Reply:
Agreed, done. This now incorporates your idea from #2109.